lib/pry/testable/repl_tester.rb

```diff
@@ -0,0 +1,155 @@
+class Pry
+  module Testable
+    #
+    # Pry::Testable::ReplTester is for super-high-level integration testing.
+    #
+    class ReplTester
+      # @api private
+      class Input
+        def initialize(tester_mailbox)
+          @tester_mailbox = tester_mailbox
+        end
+
+        def readline(prompt)
+          @tester_mailbox.push prompt
+          mailbox.pop
+        end
+
+        def mailbox
+          Thread.current[:mailbox]
+        end
+      end
+
+      require 'delegate'
+      # @api private
+      class Output < SimpleDelegator
+        def clear
+          __setobj__(StringIO.new)
+        end
+      end
+
+      #
+      # @example
+      #   Pry::Testable::ReplTester.start do |repl|
+      #     repl.enter_input '_pry_.config.prompt_name = "foo"'
+      #     expect(repl.last_prompt).to match(/foo/)
+      #   end
+      #
+      # @param [Hash] options
+      #   A hash that is passed to {Pry#initialize}.
+      #
+      # @return [void]
+      #
+      def self.start(options = {})
+        Thread.current[:mailbox] = Queue.new
+        repl_tester = new({
+          input: Input.new(Thread.current[:mailbox]),
+          output: Output.new(StringIO.new),
+          color: false
+        }.merge!(options))
+        yield repl_tester
+        repl_tester.ensure_exit
+      ensure
+        if repl_tester && repl_tester.thread && repl_tester.thread.alive?
+          repl_tester.thread.kill
+        end
+      end
+
+      attr_writer :last_prompt
+      attr_accessor :thread, :mailbox
+
+      #
+      # @return [String]
+      #   The last rendered prompt.
+      #
+      attr_reader :last_prompt
+
+      #
+      # @return [Pry]
+      #   The instance of Pry being tested.
+      #
+      attr_reader :pry
+
+      #
+      # @param [options] options
+      #   a Hash that is passed to {Pry#initialize}.
+      #
+      def initialize(options = {})
+        @pry     = Pry.new(options)
+        @repl    = Pry::REPL.new(@pry)
+        @mailbox = Thread.current[:mailbox]
+
+        @thread  = Thread.new do
+          begin
+            Thread.current[:mailbox] = Queue.new
```

Comment:
I always wondered why this is called mailbox and why it need to be stored on the thread instead of an ivar. I'd say it's a `buffer`. Anyway, can you help me understand, please?